src/newtondynamics_physics/SysNewton.cpp

```diff
@@ -76,75 +70,114 @@ using osp::Matrix3;
-        }
+        factor.m_func(pBody, bodyId, *pWorldCtx, factor.m_userData, force, torque);
     }
 
-    if (hasForce)
-    {
-        NewtonBodySetForce(pBody, force.data());
-    }
-
-    if (hasTorque)
-    {
-        NewtonBodySetTorque(pBody, torque.data());
-    }
+    NewtonBodySetForce(pBody, force.data());
+    NewtonBodySetTorque(pBody, torque.data());
 } // cb_force_torque()
 
-static void cb_set_transform(NewtonBody const* const pBody,
+void SysNewton::cb_set_transform(NewtonBody const* const pBody,
                             const dFloat* const pMatrix, int threadIndex)
 {
     // Get context from Newton World
     ACtxNwtWorld* pWorldCtx = static_cast<ACtxNwtWorld*>(
             NewtonWorldGetUserData(NewtonBodyGetWorld(pBody)));
 
     // Get Associated entity
-    uintptr_t const userData = reinterpret_cast<uintptr_t>(NewtonBodyGetUserData(pBody));
-    ActiveEnt const ent = static_cast<ActiveEnt>(userData);
+    auto const userData = reinterpret_cast<uintptr_t>(NewtonBodyGetUserData(pBody));
+    auto const bodyId = static_cast<BodyId>(userData);
+    ActiveEnt const ent = pWorldCtx->m_bodyToEnt[bodyId];
 
-    pWorldCtx->m_perThread[threadIndex].m_setTf.emplace_back(ent, pBody);
+    NewtonBodyGetMatrix(pBody, pWorldCtx->m_pTransform->get(ent).m_transform.data());
 } // cb_set_transform()
 
-} // namespace ()
+
+void SysNewton::resize_body_data(ACtxNwtWorld& rCtxWorld)
+{
+    std::size_t const capacity = rCtxWorld.m_bodyIds.capacity();
+    rCtxWorld.m_bodyPtrs    .resize(capacity);
+    rCtxWorld.m_bodyToEnt   .resize(capacity);
+    rCtxWorld.m_bodyFactors .resize(capacity);
+}
+
+NewtonCollision* SysNewton::create_primative(
+        ACtxNwtWorld&   rCtxWorld,
+        EShape          shape)
+{
+    NewtonWorld* pNwtWorld = rCtxWorld.m_world.get();
+    NewtonCollision* pCollision;
+    switch (shape)
+    {
+    case EShape::Sphere:
+        pCollision = NewtonCreateSphere(pNwtWorld, 1.0f, 0, nullptr);
+        break;
+    case EShape::Box:
+        pCollision = NewtonCreateBox(pNwtWorld, 2, 2, 2, 0, nullptr);
+        break;
+    case EShape::Cylinder:
+        pCollision = NewtonCreateCylinder(pNwtWorld, 1, 1, 2, 0, nullptr);
+        break;
+    default:
+        // TODO: support other shapes, sphere is used for now
+        pCollision = NewtonCreateSphere(pNwtWorld, 1.0f, 0, nullptr);
+        break;
+    }
+
+    return pCollision;
+}
+
+void SysNewton::orient_collision(
+        NewtonCollision const*  pCollision,
+        osp::phys::EShape       shape,
+        osp::Vector3 const&     translation,
+        osp::Matrix3 const&     rotation,
+        osp::Vector3 const&     scale)
+{
+    if (shape != EShape::Cylinder)
+    {
+        Matrix4 const matrix = Matrix4::from(rotation, translation);
+        NewtonCollisionSetMatrix(pCollision, matrix.data());
+        NewtonCollisionSetScale(pCollision, scale.x(), scale.y(), scale.z());
+    }
+    else
+    {
+        constexpr Matrix3 const rotX90deg
```

Comment:
minor: static.

Not all compilers get confused by this, but some do. I've seen assembler output where the static keyword is what makes the compiler omit the call to the constructor and do inlining properly.

That might be fixed on newer compilers though, i haven't checked in a long time.